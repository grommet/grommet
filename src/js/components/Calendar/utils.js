// Utility functions for the Calendar.
// Just what's needed to avoid having to include a dependency like momentjs.

const DAY_MILLISECONDS = 24 * 60 * 60 * 1000;

export const addDays = (date, days) => {
  const result = new Date(date.getTime() + DAY_MILLISECONDS * days);
  // Deal with crossing the daylight savings time boundary,
  // where adding a day's worth when the time is midnight results in
  // being a day off.
  let hourDelta = result.getHours() - date.getHours();
  // At this point, hourDelta is typically 0 (normal day),
  // +23 (November daylight saving), or -23 (March Daylight saving)
  // depending on which side of the switch we are on.
  // Convert so that hourDelta is either +1 or -1.
  if (hourDelta === 23) {
    hourDelta -= 24;
  } else if (hourDelta === -23) {
    hourDelta += 24;
  }
  result.setHours(result.getHours() - hourDelta);
  return result;
};

export const subtractDays = (date, days) => addDays(date, -days);

export const addMonths = (date, months) => {
  const result = new Date(date);
  const years = Math.floor((date.getMonth() + months) / 12);
  result.setFullYear(date.getFullYear() + years);
  const targetMonth = (date.getMonth() + months) % 12;
  result.setMonth(targetMonth < 0 ? 12 + targetMonth : targetMonth);
  return result;
};

export const subtractMonths = (date, months) => addMonths(date, -months);

export const startOfMonth = (date) => {
  const result = new Date(date);
  result.setDate(1);
  return result;
};

export const endOfMonth = (date) => {
  const result = addMonths(date, 1);
  result.setDate(0);
  return result;
};

export const sameDay = (date1, date2) =>
  date1.getFullYear() === date2.getFullYear() &&
  date1.getMonth() === date2.getMonth() &&
  date1.getDate() === date2.getDate();

export const sameDayOrAfter = (date1, date2) =>
  date1.getFullYear() > date2.getFullYear() ||
  (date1.getFullYear() === date2.getFullYear() &&
    (date1.getMonth() > date2.getMonth() ||
      (date1.getMonth() === date2.getMonth() &&
        date1.getDate() >= date2.getDate())));

export const sameDayOrBefore = (date1, date2) =>
  date1.getFullYear() < date2.getFullYear() ||
  (date1.getFullYear() === date2.getFullYear() &&
    (date1.getMonth() < date2.getMonth() ||
      (date1.getMonth() === date2.getMonth() &&
        date1.getDate() <= date2.getDate())));

export const daysApart = (date1, date2) =>
  Math.floor((date1.getTime() - date2.getTime()) / DAY_MILLISECONDS);

// betweenDates takes an array of two elements and checks if the
// supplied date lies between them, inclusive.
// returns 2 if exact match to one end, 1 if between, undefined otherwise
export const betweenDates = (date, dates) => {
  let result;
  if (dates) {
    const [from, to] = Array.isArray(dates)
      ? dates.map((d) => (d ? new Date(d) : undefined))
      : [dates, undefined];
    if ((from && sameDay(date, from)) || (to && sameDay(date, to))) {
      result = 2;
    } else if (
      from &&
      sameDayOrAfter(date, from) &&
      to &&
      sameDayOrBefore(date, to)
    ) {
      result = 1;
    }
  } else {
    result = 1;
  }
  return result;
};

export const getRangePosition = (date, dates) => {
  let rangePosition;
  if (dates) {
    const [from, to] = Array.isArray(dates)
      ? dates.map((d) => (d ? new Date(d) : undefined))
      : [dates, undefined];

    if (from && sameDay(date, from)) rangePosition = 'start';
    else if (to && sameDay(date, to)) rangePosition = 'end';
  }
  return rangePosition;
};

// withinDates takes an array of string dates or 2 element arrays and
// checks whether the supplied date matches any string or is between
// any dates in arrays.
// returns 2 if exact match, 1 if between, undefined otherwise
export const withinDates = (date, dates) => {
  let result;
  let rangePosition;
  if (dates) {
    if (Array.isArray(dates)) {
      dates.some((d) => {
        if (d instanceof Date) {
          if (sameDay(date, d)) {
            result = 2;
          }
        } else {
          result = betweenDates(date, d);
          rangePosition = getRangePosition(date, d);
        }
        return result;
      });
    } else if (sameDay(date, dates)) {
      result = 2;
    }
  }
  return [result, rangePosition];
};

export const handleOffset = (date) => {
  const normalizedDate = new Date(date);
  const offset = normalizedDate.getTimezoneOffset();
  const hour = normalizedDate.getHours();
  // add back offset
  normalizedDate.setHours(hour, offset < 0 ? -offset : offset);

  return normalizedDate;
};
