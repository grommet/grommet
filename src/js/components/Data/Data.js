import React, { useContext, useEffect, useMemo, useState } from 'react';
import { AnnounceContext } from '../../contexts';
import { Box } from '../Box';
import { DataFilters } from '../DataFilters';
import { DataSearch } from '../DataSearch';
import { DataSummary } from '../DataSummary';
import { DataView } from '../DataView';
import { Toolbar } from '../Toolbar';
import { DataContext } from '../../contexts/DataContext';
import { DataPropTypes } from './propTypes';
import { MessageContext } from '../../contexts/MessageContext';
import { filter } from './filter';

const defaultDefaultView = { search: '' };

const normalizeView = (viewProp, views) =>
  (typeof viewProp === 'string' && views?.find((v) => v.name === viewProp)) ||
  (typeof viewProp === 'object' && viewProp);

export const Data = ({
  children,
  data: dataProp = [],
  defaultView = defaultDefaultView,
  filteredTotal,
  id = 'data',
  messages,
  onView,
  properties,
  toolbar,
  total,
  view: viewProp,
  views,
  ...rest
}) => {
  const [view, setView] = useState(
    normalizeView(viewProp || defaultView, views),
  );
  useEffect(
    () => setView(normalizeView(viewProp || defaultView, views)),
    [defaultView, viewProp, views],
  );
  const [toolbarKeys, setToolbarKeys] = useState([]);

  const result = useMemo(() => {
    if (onView)
      // caller is filtering
      return {
        data: dataProp,
        total,
        filteredTotal: filteredTotal ?? dataProp?.length ?? 0,
      };
    return filter(dataProp, view, properties);
  }, [dataProp, filteredTotal, onView, properties, total, view]);

  // used by DataFilters to determine if badge should appear on Filter button
  const [filtersCleared, setFiltersCleared] = useState(true);

  const [selected, setSelected] = useState([]);

  const announce = useContext(AnnounceContext);
  const { format } = useContext(MessageContext);
  // Announce to screen readers when search or filters are
  // applied and affect the underlying result set
  useEffect(() => {
    let messageId;
    if (result.total !== result.filteredTotal) {
      if (result.filteredTotal === 1) messageId = 'dataSummary.filteredSingle';
      else messageId = 'dataSummary.filtered';
    } else if (result.total === 1) messageId = 'dataSummary.totalSingle';
    else messageId = 'dataSummary.total';

    // helps account for cases like 0 results of 1 item
    const items = format({
      id: result.total === 1 ? 'dataSummary.itemsSingle' : 'dataSummary.items',
      messages: messages?.dataSummary,
    });

    announce(
      `${format({
        id: messageId,
        messages: messages?.dataSummary,
        values: {
          filteredTotal: result.filteredTotal,
          total: result.total,
          items,
        },
      })}${
        selected > 0
          ? `, ${format({
              id: 'dataSummary.selected',
              messages: messages?.dataSummary,
              values: {
                selected,
              },
            })}`
          : ''
      }`,
    );
  }, [
    announce,
    format,
    messages?.dataSummary,
    result.filteredTotal,
    result.total,
    selected,
  ]);

  // what we use for DataContext value
  const contextValue = useMemo(() => {
    const value = {
      id,
      messages,
      properties,
      filtersCleared,
      setFiltersCleared,
      selected,
      setSelected,
      view,
      views,
      ...result,
    };

    value.clearFilters = () => {
      const nextView = { ...view };
      delete nextView.properties;
      delete nextView.page;
      // by clearing the properties from a view, it is no
      // longer reflecting the view
      delete nextView.name;
      delete nextView.view;
      setFiltersCleared(true);
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.onView = (nextView) => {
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.addToolbarKey = (key) => {
      setToolbarKeys((prevKeys) => {
        if (prevKeys.includes(key)) return prevKeys;
        return [...prevKeys, key];
      });
    };
    value.toolbarKeys = toolbarKeys;

    return value;
  }, [
    id,
    messages,
    filtersCleared,
    onView,
    properties,
    result,
    selected,
    toolbarKeys,
    view,
    views,
  ]);

  let toolbarContent;
  if (toolbar) {
    toolbarContent = [
      // TO DO theme object
      <Toolbar key="toolbar" gap="medium">
        <Toolbar>
          {(toolbar === true || toolbar === 'search') && <DataSearch />}
          {(toolbar === true || toolbar === 'filters') && <DataFilters layer />}
        </Toolbar>
        {(toolbar === true || toolbar === 'view') && <DataView />}
      </Toolbar>,
      <DataSummary key="summary" />,
    ];
  }

  return (
    <DataContext.Provider value={contextValue}>
      <Box id={id} flex={false} {...rest}>
        {toolbarContent}
        {children}
      </Box>
    </DataContext.Provider>
  );
};

Data.propTypes = DataPropTypes;
