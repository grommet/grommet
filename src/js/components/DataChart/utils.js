import { round } from '../Chart';

export const points = [
  'circle',
  'diamond',
  'square',
  'star',
  'triangle',
  'triangleDown',
];

export const heightYGranularity = {
  xxsmall: { fine: 2, medium: 2 },
  xsmall: { fine: 3, medium: 2 },
  small: { fine: 5, medium: 3 },
  medium: { fine: 7, medium: 5 },
  large: { fine: 9, medium: 5 },
  xlarge: { fine: 11, medium: 5 },
};

export const halfPad = (theme) => ({
  xlarge: theme.dataChart?.halfPad?.xlarge || 'large',
  large: theme.dataChart?.halfPad?.large || 'medium',
  medium: theme.dataChart?.halfPad?.medium || 'small',
  small: theme.dataChart?.halfPad?.small || 'xsmall',
  xsmall: theme.dataChart?.halfPad?.xsmall || 'xxsmall',
});

export const doublePad = (theme) => ({
  large: theme.dataChart?.doublePad?.large || 'xlarge',
  medium: theme.dataChart?.doublePad?.medium || 'large',
  small: theme.dataChart?.doublePad?.small || 'medium',
  xsmall: theme.dataChart?.doublePad?.xsmall || 'small',
  xxsmall: theme.dataChart?.doublePad?.xxsmall || 'xsmall',
});

const getOrderedSizes = (theme) =>
  theme?.dataChart?.orderedSizes || [
    'xlarge',
    'large',
    'medium',
    'small',
    'xsmall',
    'xxsmall',
    'hair',
  ];

export const showInUnits = (content, maxValue) => {
  let divideBy;
  let unit;
  let newContent = content;
  if (maxValue > 10000000) {
    divideBy = 1000000;
    unit = 'M';
  } else if (maxValue > 10000) {
    divideBy = 1000;
    unit = 'K';
  }
  if (divideBy) newContent = round(newContent / divideBy, 0);
  if (unit) newContent = `${newContent}${unit}`;
  return newContent;
};

export const largestSize = (theme, size1, size2) => {
  const orderedSizes = getOrderedSizes(theme);
  if (size1 && !size2) return size1;
  if (size2 && !size1) return size2;
  if (orderedSizes.indexOf(size1) < orderedSizes.indexOf(size2)) return size1;
  return size2;
};

export const createDateFormat = (firstValue, lastValue, full) => {
  let dateFormat;
  const startDate = new Date(firstValue);
  const endDate = new Date(lastValue);
  if (
    // check for valid dates, this is the fastest way
    !Number.isNaN(startDate.getTime()) &&
    !Number.isNaN(endDate.getTime())
  ) {
    const delta = Math.abs(endDate - startDate);
    let options;
    if (delta < 60000)
      // less than 1 minute
      options = full
        ? {
            hour: '2-digit',
            minute: '2-digit',
            second: '2-digit',
            day: undefined,
          }
        : { second: '2-digit', day: undefined };
    else if (delta < 3600000)
      // less than 1 hour
      options = full
        ? { hour: 'numeric', minute: '2-digit', day: undefined }
        : { minute: '2-digit', day: undefined };
    else if (delta < 86400000)
      // less than 1 day
      options = { hour: 'numeric' };
    else if (delta < 2592000000)
      // less than 30 days
      options = {
        month: full ? 'short' : 'numeric',
        day: 'numeric',
      };
    else if (delta < 31557600000)
      // less than 1 year
      options = { month: full ? 'long' : 'short' };
    // 1 year or more
    else options = { year: 'numeric' };
    if (options)
      dateFormat = new Intl.DateTimeFormat(undefined, options).format;
  }
  return dateFormat;
};

export const minimum = (value1, value2) =>
  value1 !== undefined ? Math.min(value1, value2) : value2;

export const maximum = (value1, value2) =>
  value1 !== undefined ? Math.max(value1, value2) : value2;
