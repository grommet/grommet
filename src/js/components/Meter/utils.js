import { normalizeColor } from '../../utils';

export const strokeProps = (color, theme) => {
  const result = {};
  if (color) {
    if (typeof color === 'object') {
      result.stroke = normalizeColor(color.color, theme);
      if (color.opacity) {
        result.strokeOpacity = `${
          color.opacity === true
            ? theme.global.opacity.medium
            : theme.global.opacity[color.opacity] || color.opacity
        }`;
      }
    } else {
      result.stroke = normalizeColor(color, theme);
    }
  }
  return result;
};

export const fillProps = (color, theme) => {
  const result = {};
  if (color) {
    if (typeof color === 'object') {
      result.fill = normalizeColor(color.color, theme);
      // if (color.opacity) {
      //   result.strokeOpacity = `${
      //     color.opacity === true
      //       ? theme.global.opacity.medium
      //       : theme.global.opacity[color.opacity] || color.opacity
      //   }`;
      // }
    } else {
      result.fill = normalizeColor(color, theme);
    }
  }
  return result;
};

const neutralExp = /^neutral-\d+/;

export const defaultColor = (index, theme, valuesLength) => {
  if (index === valuesLength - 1 && theme.meter.color) {
    return theme.meter.color;
  }
  // We want the last value to have the first color
  const colorIndex = valuesLength - index - 1;
  if (theme.meter && theme.meter.colors) {
    const colors =
      theme.meter.colors[theme.dark ? 'dark' : 'light'] || theme.meter.colors;
    return colors[colorIndex % colors.length];
  }
  const colors = Object.keys(theme.global.colors).filter(n =>
    n.match(/^graph-[0-9]$/),
  );
  if (colors.length > 0) {
    return colors[colorIndex % colors.length];
  }
  // Deprecate using "neutral-*" color names. Remove eventually.
  const neutralColors = Object.keys(theme.global.colors).filter(k =>
    neutralExp.test(k),
  );
  return neutralColors[colorIndex % neutralColors.length];
};
