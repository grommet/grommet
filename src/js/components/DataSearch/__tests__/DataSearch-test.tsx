import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { Data } from '../../Data';
import { DataFilters } from '../../DataFilters';
import { Grommet } from '../../Grommet';
import { DataSearch } from '..';
import { createPortal, expectPortal } from '../../../utils/portal';

const data = [{ name: 'a' }, { name: 'b' }];

describe('DataSearch', () => {
  window.scrollTo = jest.fn();
  beforeEach(createPortal);

  test('renders', () => {
    const { container } = render(
      <Grommet>
        <Data data={data}>
          <DataFilters>
            <DataSearch />
          </DataFilters>
        </Data>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('no Data', () => {
    const { container } = render(
      <Grommet>
        <DataSearch />
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('drop', () => {
    jest.useFakeTimers();

    const { container, getByRole } = render(
      <Grommet>
        <Data id="test-data" data={data}>
          <DataSearch drop />
        </Data>
      </Grommet>,
    );

    expect(getByRole('button', { name: 'Open search' })).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();

    fireEvent.click(getByRole('button', { name: 'Open search' }));
    // advance timers so drop can open
    act(() => jest.advanceTimersByTime(200));

    // snapshot on drop
    expectPortal('test-data--search-control').toMatchSnapshot();
  });

  test('theme data drop pad', () => {
    const customTheme = {
      data: {
        drop: {
          pad: 'large',
        },
      },
    };

    const { asFragment } = render(
      <Grommet theme={customTheme}>
        <Data data={data}>
          <DataFilters>
            <DataSearch drop />
          </DataFilters>
        </Data>
      </Grommet>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

test('enter', async () => {
  jest.useFakeTimers();
  const onView = jest.fn();
  const { getByRole } = render(
    <Grommet>
      <Data data={data} onView={onView}>
        <DataSearch data-testid="input_submit" updateOn="submit" />
      </Data>
    </Grommet>,
  );
  const searchbox = getByRole('searchbox');
  expect(searchbox).toBeTruthy();

  fireEvent.change(searchbox, { target: { value: 'one' } });
  act(() => jest.advanceTimersByTime(300));

  fireEvent.keyDown(searchbox, { key: 'enter', keyCode: 13 });

  expect(onView).toHaveBeenNthCalledWith(
    1,
    expect.objectContaining({
      search: 'one',
    }),
  );
}, 20000);

test('change', async () => {
  jest.useFakeTimers();
  const onView = jest.fn();
  const { getByRole } = render(
    <Grommet>
      <Data data={data} onView={onView}>
        <DataSearch data-testid="input_change" updateOn="change" />
      </Data>
    </Grommet>,
  );
  const searchbox = getByRole('searchbox');
  expect(searchbox).toBeTruthy();

  fireEvent.change(searchbox, { target: { value: 'two' } });
  act(() => jest.advanceTimersByTime(300));

  expect(onView).toHaveBeenNthCalledWith(
    1,
    expect.objectContaining({
      search: 'two',
    }),
  );
}, 20000);
