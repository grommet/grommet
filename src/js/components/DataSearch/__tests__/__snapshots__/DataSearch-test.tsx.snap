// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSearch drop 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c3 g {
  fill: inherit;
  stroke: inherit;
}

.c3 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c3 *[stroke*='#'],
.c3 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c3 *[fill-rule],
.c3 *[FILL-RULE],
.c3 *[fill*='#'],
.c3 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c2 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  line-height: 0;
  padding: 12px;
}

.c2:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c2:focus >circle,
.c2:focus >ellipse,
.c2:focus >line,
.c2:focus >path,
.c2:focus >polygon,
.c2:focus >polyline,
.c2:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c2:focus ::-moz-focus-inner {
  border: 0;
}

.c2:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c2:focus:not(:focus-visible) >circle,
.c2:focus:not(:focus-visible) >ellipse,
.c2:focus:not(:focus-visible) >line,
.c2:focus:not(:focus-visible) >path,
.c2:focus:not(:focus-visible) >polygon,
.c2:focus:not(:focus-visible) >polyline,
.c2:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c2:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  class="c0"
>
  <div
    class="c1"
    id="test-data"
  >
    <button
      aria-label="Open search"
      class="c2"
      id="test-data--search-control"
      type="button"
    >
      <svg
        aria-label="Search"
        class="c3"
        viewBox="0 0 24 24"
      >
        <path
          d="m15 15 7 7-7-7zm-5.5 2a7.5 7.5 0 1 0 0-15 7.5 7.5 0 0 0 0 15z"
          fill="none"
          stroke="#000"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`DataSearch drop 2`] = `
<button
  aria-label="Open search"
  class="StyledButton-sc-323bzc-0 geHpTD"
  data-g-tabindex="none"
  id="test-data--search-control"
  tabindex="-1"
  type="button"
>
  <svg
    aria-label="Search"
    class="StyledIcon-sc-ofa7kd-0 bqRbRY"
    viewBox="0 0 24 24"
  >
    <path
      d="m15 15 7 7-7-7zm-5.5 2a7.5 7.5 0 1 0 0-15 7.5 7.5 0 0 0 0 15z"
      fill="none"
      stroke="#000"
      stroke-width="2"
    />
  </svg>
</button>
`;

exports[`DataSearch drop 3`] = `
"@media only screen and (max-width: 768px) {
  .lAiqC {
    margin-bottom: 6px;
  }
}
@media only screen and (max-width: 768px) {
  .eYrCJw {
    border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  }
}
@media only screen and (max-width: 768px) {
  .eYqvdH {
    margin-top: 12px;
  }
}"
`;

exports[`DataSearch no Data 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c4 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c4 g {
  fill: inherit;
  stroke: inherit;
}

.c4 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c4 *[stroke*='#'],
.c4 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c4 *[fill-rule],
.c4 *[FILL-RULE],
.c4 *[fill*='#'],
.c4 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c5 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0, 0, 0, 0.33);
  border-radius: 4px;
  padding-left: 48px;
}

.c5:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c5:focus >circle,
.c5:focus >ellipse,
.c5:focus >line,
.c5:focus >path,
.c5:focus >polygon,
.c5:focus >polyline,
.c5:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c5:focus ::-moz-focus-inner {
  border: 0;
}

.c5::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c5::-moz-placeholder {
  color: #AAAAAA;
}

.c5:-ms-input-placeholder {
  color: #AAAAAA;
}

.c5 ::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c5::-moz-focus-inner {
  border: none;
  outline: none;
}

.c5:-moz-placeholder,
.c5::-moz-placeholder {
  opacity: 1;
}

.c2 {
  position: relative;
  width: 100%;
}

.c3 {
  position: absolute;
  display: flex;
  justify: center;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  left: 12px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <svg
          aria-label="Search"
          class="c4"
          viewBox="0 0 24 24"
        >
          <path
            d="m15 15 7 7-7-7zm-5.5 2a7.5 7.5 0 1 0 0-15 7.5 7.5 0 0 0 0 15z"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
      <input
        aria-label="Search"
        autocomplete="off"
        class="c5"
        id="undefined--search"
        name="_search"
        type="search"
      />
    </div>
  </div>
</div>
`;

exports[`DataSearch renders 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-bottom: 12px;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c7 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c12 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c10 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c10 g {
  fill: inherit;
  stroke: inherit;
}

.c10 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c10 *[stroke*='#'],
.c10 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c10 *[fill-rule],
.c10 *[FILL-RULE],
.c10 *[fill*='#'],
.c10 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c6 {
  margin-left: 12px;
  margin-right: 12px;
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c13 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c13:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c13:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c13:focus >circle,
.c13:focus >ellipse,
.c13:focus >line,
.c13:focus >path,
.c13:focus >polygon,
.c13:focus >polyline,
.c13:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c13:focus ::-moz-focus-inner {
  border: 0;
}

.c13:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c13:focus:not(:focus-visible) >circle,
.c13:focus:not(:focus-visible) >ellipse,
.c13:focus:not(:focus-visible) >line,
.c13:focus:not(:focus-visible) >path,
.c13:focus:not(:focus-visible) >polygon,
.c13:focus:not(:focus-visible) >polyline,
.c13:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c13:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c2 {
  max-width: 100%;
}

.c11 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0, 0, 0, 0.33);
  border-radius: 4px;
  outline: none;
  border: none;
  padding-left: 48px;
}

.c11::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c11::-moz-placeholder {
  color: #AAAAAA;
}

.c11:-ms-input-placeholder {
  color: #AAAAAA;
}

.c11 ::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c11::-moz-focus-inner {
  border: none;
  outline: none;
}

.c11:-moz-placeholder,
.c11::-moz-placeholder {
  opacity: 1;
}

.c8 {
  position: relative;
  width: 100%;
}

.c9 {
  position: absolute;
  display: flex;
  justify: center;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  left: 12px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {
  .c5 {
    margin-bottom: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c7 {
    border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  }
}

@media only screen and (max-width: 768px) {
  .c12 {
    margin-top: 12px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    id="data"
  >
    <form
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c1"
          >
            <div
              class="c5 "
            >
              <label
                class="c6"
                for="data--search"
              >
                Search
              </label>
              <div
                class="c7 "
              >
                <div
                  class="c8"
                >
                  <div
                    class="c9"
                  >
                    <svg
                      aria-label="Search"
                      class="c10"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m15 15 7 7-7-7zm-5.5 2a7.5 7.5 0 1 0 0-15 7.5 7.5 0 0 0 0 15z"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </div>
                  <input
                    aria-label="Search"
                    autocomplete="off"
                    class="c11"
                    id="data--search"
                    name="_search"
                    type="search"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer
          class="c12"
        >
          <button
            class="c13"
            type="submit"
          >
            Apply filters
          </button>
        </footer>
      </div>
    </form>
  </div>
</div>
`;
